File: 46. Permutations/Fixed/permutations.kt
fun exponented (base: Int, exponent: Int): Int{
	var out: Int = 1;
	
	for (i in 0 until exponent){
		out *= base;
	}
	
	return (out);
}

fun factorial (number: Int): Int{
	var out: Int = 1;
	
	for (i in number downTo 1){
		out *= i;
	}
	
	return (out);
}

fun number_from_list (list: Array<Int>): Int{
	var out: Int = 0;
	val len_list: Int = list.size;
	val base: Int = 10;
	
	for (i in 0 until len_list){
		out += (list[i] * exponented (base, (len_list - 1 - i)));
	}
	
	return (out);
}

fun number_to_list(number: Int, len_list: Int): Array<Int>{
	val base: Int = 10;
	var running_number: Int = number;
	var out: Array<Int> = Array (len_list) {0};
	var out_cursor = len_list - 1;
	
	while ((running_number > 0) && (out_cursor >= 0)){
		var modulo = running_number % base;
		running_number -= modulo;
		running_number /= base;
		out[out_cursor] = modulo;
		out_cursor -= 1;
	}
	
	return (out);
}

fun list_comparison (list1: Array<Int>, list2: Array<Int>): Boolean{
	val len_list1: Int = list1.size;
	val len_list2: Int = list2.size;
	
	if (len_list1 != len_list2){
		return (false);
	}
	
	var list1_bool_map: Array<Boolean> = Array(len_list1) {false};
	var list2_bool_map: Array<Boolean> = Array(len_list2) {false};
	
	for (i in 0 until len_list1){
		for (j in 0 until len_list2){
			if ((list1_bool_map[i] == false) && (list2_bool_map[j] == false)){
				if (list1[i] == list2[j]){
					list1_bool_map[i] = true;
					list2_bool_map[j] = true;
					break;
				}
			}
		}
	}
	
	var bool_counter: Int = 0;
	
	for (i in 0 until len_list1){
		if (list1_bool_map[i] && list2_bool_map[i]){
			bool_counter += 1;
		}
	}
	
	return (len_list1 == bool_counter);
}

fun appears (element: Int, list: Array<Int>): Int {
	var out: Int = -1;
	
	for (i in 0 until list.size){
		if (element == list[i]){
			out += 1;
		}
	}
	
	return (out);
}

fun main (){
	val nums: Array<Int> = arrayOf(1,2,3,4);
	val len_nums: Int = nums.size;
	
	nums.sort();
	val lowest_value_nums_number = number_from_list(nums);
	
	nums.reverse();
	val highest_value_nums_number = number_from_list(nums);
	
	print("The lowest number from the elements is:\t");
	println(lowest_value_nums_number);
	print("The highest number from the elements is:\t");
	println(highest_value_nums_number);
	
	var sample_size: Int = factorial(len_nums);
	var unique: Int = 0;
	
	for (i in 0 until len_nums){
		var occ_code: Boolean = false;
		
		if (i == 0){
			occ_code = true;
		} else {
			if (nums[i - 1] != nums[i]){
				occ_code = true;
			}
		}
		
		if (occ_code){
			var occurences = appears(nums[i], nums)
			if (occurences == 0){
				unique += 1;
			}
			
			if (occurences > 0){
				unique += 1;
				sample_size /= factorial(occurences + 1);
			}
		}
	}
	
	print("The number of unique elements are:\t");
	println(unique);
	print("The sample size of permutations expected is:\t");
	println(sample_size);
	
	var permutations: Array<Array<Int>> = Array(sample_size) {Array(len_nums) {0}};
	var running_number: Int = lowest_value_nums_number;
	var permutations_cursor: Int = 0;
	
	while (running_number <= highest_value_nums_number){
		var temp_list: Array<Int> = number_to_list(running_number, len_nums);
		
		if (list_comparison(temp_list, nums)){
			for (i in 0 until len_nums){
				permutations[permutations_cursor][i] = temp_list[i];
			}
			permutations_cursor += 1;
		}
		
		running_number += 1;
	}
	
	for (i in 0 until permutations_cursor){
		for (j in 0 until len_nums){
			print(permutations[i][j]);
			
			if (j != (len_nums - 1)){
				print("\t");
			}
		}
		
		println();
	}
}

File: 46. Permutations/Fixed/permutations.cs
using System;

namespace permutations_ns
{
	class permutations_c
	{
		static int exponented (int radix, int exponent){
			int nje = 1;
			
			for (int i = 0; i < exponent; i++){
				nje *= radix;
			}
			
			return (nje);
		}
		
		static int factorial (int number){
			int nje = 1;
			
			for (int i = number; i > 0; i--){
				nje *= i;
			}
			
			return (nje);
		}
		
		static int number_from_list (int [] list){
			int len_list = list.Length;
			int nje = 0;
			int radix = 10;
			
			for (int i = 0; i < len_list; i++){
				nje += (list[i] * exponented(radix, (len_list - 1 - i)));
			}
			
			return (nje);
		}
		
		static int [] number_to_list (int number, int len_list){
			int [] nje = new int [len_list];
			
			for (int i = 0; i < len_list; i++){
				nje[i] = 0;
			}
			
			int radix = 10;
			int running_number = number;
			int nje_cursor = len_list - 1;
			
			while ((running_number > 0) && (nje_cursor >= 0)){
				int modulo = running_number % radix;
				running_number -= modulo;
				running_number /= radix;
				nje[nje_cursor] = modulo;
				nje_cursor--;
			}
			
			return (nje);
		}
		
		static bool list_comparison (int [] list1, int [] list2){
			int len_list1 = list1.Length;
			int len_list2 = list2.Length;
			
			if (len_list1 != len_list2){
				return (false);
			}
			
			bool [] list1_bool_map = new bool[len_list1];
			bool [] list2_bool_map = new bool[len_list1];
			
			for (int i = 0; i < len_list1; i++){
				list1_bool_map[i] = false;
				list2_bool_map[i] = false;
			}
			
			for (int i = 0; i < len_list1; i++){
				for (int j = 0; j < len_list1; j++){
					if ((list1_bool_map[i] == false) && (list2_bool_map[j] == false)){
						if (list1[i] == list2[j]){
							list1_bool_map[i] = true;
							list2_bool_map[j] = true;
							break;
						}
					}
				}
			}
			
			int bool_counter = 0;
			
			for (int i = 0; i < len_list1; i++){
				if (list1_bool_map[i] && list2_bool_map[i]){
					bool_counter++;
				}
			}
			
			return (len_list1 == bool_counter);
		}
		
		static int appears (int element, int [] list){
			int len_list = list.Length;
			int nje = -1;
			
			for (int i = 0; i < len_list; i++){
				if (element == list[i]){
					nje += 1;
				}
			}
			return (nje);
		}
		
		//reverse
		//compare
		
		static void Main (String [] args){
			int [] nums = {1,2,3,4};
			int len_nums = nums.Length;
			
			//sort
			Array.Sort(nums);
			
			int lowest_value_nums_number = number_from_list(nums);
			//reverse
			Array.Reverse(nums);
			int highest_value_nums_number = number_from_list(nums);
			
			Console.Write("The lowest number from the elements is:\t");
			Console.WriteLine(lowest_value_nums_number);
			Console.Write("The highest number from the elements is:\t");
			Console.WriteLine(highest_value_nums_number);
			
			int sample_size = factorial(len_nums);
			int unique = 0;
			
			for (int i = 0; i < len_nums; i++){
				bool occ_code = false;
				
				if (i == 0){
					occ_code = true;
				} else {
					if (nums[i - 1] != nums[i]){
						occ_code = true;
					}
				}
				
				if (occ_code){
					int occurences = appears(nums[i], nums);
					if (occurences == 0){
						unique++;
					}
					
					if (occurences > 0){
						unique++;
						sample_size /= factorial(occurences + 1);
					}
				}
			}
			
			Console.Write("The number of unique elements are:\t");
			Console.WriteLine(unique);
			Console.Write("The sample size of permutations expected is:\t");
			Console.WriteLine(sample_size);
			Console.WriteLine("\n\n\n");
			
			int [,] permutations = new int [sample_size, len_nums];
			int running_number = lowest_value_nums_number;
			int permutations_cursor = 0;
			
			while (running_number <= highest_value_nums_number){
				int [] temp_array = number_to_list(running_number, len_nums);
				
				if (list_comparison(temp_array, nums)){
					for (int i = 0; i < len_nums; i++){
						permutations[permutations_cursor, i] = temp_array[i];
					}
					permutations_cursor++;
				}
				
				running_number++;
			}
			
			for (int i = 0; i < permutations_cursor; i++){
				for (int j = 0; j < len_nums; j++){
					Console.Write(permutations[i, j]);
					/*
					if (j != (len_nums - 1)){
						Console.Write("\t");
					}
					*/
				}
				Console.WriteLine();
			}
			Console.WriteLine("\n\n\n");
		}
	}
}
